File: Sources/SnipKeyCore/SnippetTriggerRules.swift
import Foundation

public enum SnippetTriggerRules {
    public static let defaultBase = "key"

    public enum ValidationError: Equatable {
        case empty
        case invalidCharacters
        case duplicate
    }

    public static func sanitize(_ trigger: String) -> String {
        String(trigger.filter(isAllowedCharacter))
    }

    public static func validationError(
        for trigger: String,
        existingTriggers: [String]
    ) -> ValidationError? {
        guard trigger.isEmpty == false else {
            return .empty
        }

        guard sanitize(trigger) == trigger else {
            return .invalidCharacters
        }

        let normalizedTrigger = trigger.lowercased()
        if existingTriggers.contains(where: { $0.lowercased() == normalizedTrigger }) {
            return .duplicate
        }

        return nil
    }

    public static func nextAvailableTrigger(
        existingTriggers: [String],
        base: String = defaultBase
    ) -> String {
        normalizedTrigger(from: base, existingTriggers: existingTriggers)
    }

    public static func normalizedTrigger(
        from trigger: String,
        existingTriggers: [String],
        fallbackBase: String = defaultBase
    ) -> String {
        let sanitized = sanitize(trigger)
        let base = sanitized.isEmpty ? sanitize(fallbackBase) : sanitized
        let resolvedBase = base.isEmpty ? defaultBase : base
        let normalizedExisting = Set(existingTriggers.map { $0.lowercased() })

        if normalizedExisting.contains(resolvedBase.lowercased()) == false {
            return resolvedBase
        }

        let separator = resolvedBase.hasSuffix("_") ? "" : "_"
        var suffix = 2
        while true {
            let candidate = resolvedBase + separator + String(suffix)
            if normalizedExisting.contains(candidate.lowercased()) == false {
                return candidate
            }
            suffix += 1
        }
    }

    public static func isAllowedCharacter(_ character: Character) -> Bool {
        if character == "_" {
            return true
        }

        guard character.unicodeScalars.count == 1,
              let scalar = character.unicodeScalars.first,
              scalar.isASCII else {
            return false
        }

        return CharacterSet.alphanumerics.contains(scalar)
    }
}

File: Windows/SnipKey.Windows/Core/SnippetTriggerRules.cs
namespace SnipKey.WinApp.Core;

public static class SnippetTriggerRules
{
    public const string DefaultBase = "key";

    public enum ValidationError
    {
        Empty,
        InvalidCharacters,
        Duplicate
    }

    public static string Sanitize(string trigger)
    {
        return new string(trigger.Where(IsAllowedCharacter).ToArray());
    }

    public static ValidationError? Validate(string trigger, IEnumerable<string> existingTriggers)
    {
        if (string.IsNullOrEmpty(trigger))
        {
            return ValidationError.Empty;
        }

        if (Sanitize(trigger) != trigger)
        {
            return ValidationError.InvalidCharacters;
        }

        var normalizedTrigger = trigger.ToLowerInvariant();
        if (existingTriggers.Any(existingTrigger => string.Equals(existingTrigger, normalizedTrigger, StringComparison.OrdinalIgnoreCase)))
        {
            return ValidationError.Duplicate;
        }

        return null;
    }

    public static string NextAvailableTrigger(IEnumerable<string> existingTriggers, string baseTrigger = DefaultBase)
    {
        return NormalizedTrigger(baseTrigger, existingTriggers);
    }

    public static string NormalizedTrigger(string trigger, IEnumerable<string> existingTriggers, string fallbackBase = DefaultBase)
    {
        var sanitized = Sanitize(trigger);
        var baseValue = string.IsNullOrEmpty(sanitized) ? Sanitize(fallbackBase) : sanitized;
        var resolvedBase = string.IsNullOrEmpty(baseValue) ? DefaultBase : baseValue;
        var normalizedExisting = existingTriggers.Select(existingTrigger => existingTrigger.ToLowerInvariant()).ToHashSet();

        if (!normalizedExisting.Contains(resolvedBase.ToLowerInvariant()))
        {
            return resolvedBase;
        }

        var separator = resolvedBase.EndsWith('_') ? string.Empty : "_";
        var suffix = 2;
        while (true)
        {
            var candidate = resolvedBase + separator + suffix;
            if (!normalizedExisting.Contains(candidate.ToLowerInvariant()))
            {
                return candidate;
            }

            suffix += 1;
        }
    }

    public static bool IsAllowedCharacter(char character)
    {
        return character == '_' || (character <= 127 && char.IsLetterOrDigit(character));
    }
}
